File: tools/sdk-ai-bots/azure-sdk-qa-bot-knowledge-sync/src/services/AppConfig.ts
import { DefaultAzureCredential } from '@azure/identity';
import { AppConfigurationClient } from '@azure/app-configuration';
import * as dotenv from 'dotenv';

/**
 * Load environment variables from .env file if it exists
 */
function loadEnvFile(): void {
    try {
        // Go to project root (3 levels up from dist/src/services when compiled)
        const path = require('path');
        const projectRoot = path.resolve(__dirname, '../../..');
        const envPath = path.join(projectRoot, '.env');
        
        dotenv.config({ path: envPath });
        console.log(`Loaded environment variables from .env file at: ${envPath}`);
    } catch (error) {
        console.log(`No .env file found or error loading it: ${error}`);
    }
}

/**
 * Initialize configuration by loading from local .env file first, then Azure App Configuration
 */
export async function initConfiguration(): Promise<void> {
    
    // Load .env file first
    loadEnvFile();

    // Get the endpoint from environment variable (could be from .env or already set)
    const endpoint = process.env.AZURE_APPCONFIG_ENDPOINT;
    
    if (!endpoint) {
        throw new Error('AZURE_APPCONFIG_ENDPOINT environment variable is required');
    }

    try {
        console.log('Loading configuration from Azure App Configuration...');
        
        // Create a credential using DefaultAzureCredential
        const credential = new DefaultAzureCredential();

        // Create the App Configuration client
        const client = new AppConfigurationClient(endpoint, credential);

        // Load all configuration settings
        const settings = client.listConfigurationSettings();

        // Iterate through all settings and set them as environment variables
        // Only set if not already defined (giving priority to .env file)
        for await (const setting of settings) {
            if (setting.key && setting.value !== undefined) {
                if (!process.env[setting.key]) {
                    process.env[setting.key] = setting.value;
                    console.log(`Set ${setting.key} from App Configuration`);
                } else {
                    console.log(`Skipping ${setting.key} - already set in environment (likely from .env file)`);
                }
            }
        }

        console.log('Successfully loaded configuration from Azure App Configuration');

    } catch (error) {
        console.error('Failed to load configuration from Azure App Configuration:', error);
        throw error;
    }
}

File: tools/sdk-ai-bots/azure-sdk-qa-bot-knowledge-graph-sync/src/services/app_config.py
"""Azure App Configuration loader.

Loads settings from Azure App Configuration and sets them as environment
variables (only if not already set, giving priority to .env/local values).
"""

from __future__ import annotations

import logging
import os
from pathlib import Path

from dotenv import load_dotenv

logger = logging.getLogger(__name__)


def _load_env_file() -> None:
    """Load .env file from project root."""
    project_root = Path(__file__).resolve().parent.parent.parent
    env_path = project_root / ".env"
    if env_path.exists():
        load_dotenv(env_path)
        logger.info("Loaded environment variables from %s", env_path)
    else:
        logger.debug("No .env file at %s", env_path)


async def init_configuration() -> None:
    """Initialize configuration from .env then Azure App Configuration.

    Settings from App Configuration only override env vars that are not
    already set (local .env takes priority).
    """
    _load_env_file()

    endpoint = os.environ.get("AZURE_APPCONFIG_ENDPOINT")
    if not endpoint:
        logger.warning("AZURE_APPCONFIG_ENDPOINT not set; skipping App Configuration")
        return

    from azure.appconfiguration.aio import AzureAppConfigurationClient
    from azure.identity.aio import (
        AzureCliCredential,
        ChainedTokenCredential,
        ManagedIdentityCredential,
        WorkloadIdentityCredential,
    )

    logger.info("Loading configuration from Azure App Configuration...")

    credential = ChainedTokenCredential(
        ManagedIdentityCredential(),
        AzureCliCredential(),
        WorkloadIdentityCredential(),
    )

    client = AzureAppConfigurationClient(endpoint, credential=credential)

    try:
        async for setting in client.list_configuration_settings():
            if setting.key and setting.value is not None:
                if not os.environ.get(setting.key):
                    os.environ[setting.key] = setting.value
                    logger.info("Set %s from App Configuration", setting.key)
                else:
                    logger.debug("Skipping %s — already set", setting.key)
    finally:
        await client.close()
        await credential.close()

    logger.info("Successfully loaded configuration from Azure App Configuration")
